---
title: "Create OmicSignature"
output: rmarkdown::html_vignette
author: "Vanessa Mengze Li"
date: "08/15/2024"
vignette: >
  %\VignetteIndexEntry{Create OmicSignature}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r message=FALSE, warning=FALSE}
devtools::load_all(".")
library(dplyr)
```
<font color = "#000000">

## 1. Cheatsheet  
An `OmicSignature` object contains three parts: \

- **metadata**, a list. \
  Required fields: \
  "**signature_name**", "**organism**", "**direction_type**", "**assay_type**", "**phenotype**". \
  Recommended optional fields, if applicable: \
	"platform", "sample_type", "description", "covariates", "score_cutoff", "adj_p_cutoff". \

- **signature**, a data frame. \
  Required columns: "**id**", "**symbol**" \
  Required for bi-directional and categorical signatures: "**direction**" \
  Recommended optional column, if applicable: "score" \

- **difexp** (optional), a data frame of differential expression analysis results. \
  Required columns: "**id**", "**symbol**", \
  Require at least one of the following: "**p_value**", "**q_value**", or "**adj_p**". \

Create the object: \
```
OmicObj <- OmicSignature$new(
  metadata = metadata,
  signature = signature,
  difexp = difexp
)
```

## 2. Create an `OmicSignature` Object Step-by-Step

The example provided below is from an experiment for Myc gene reduce in mice. Signatures were extracted by comparing the liver of treatment and control when mice is 24-month old. To save space, only the result of 1000 genes are included. \

This is a bi-directional signature example, which contains up and down regulated features (genes). \

### 2.1. Metadata
A list with the following required fields: \
"**signature_name**", "**organism**", "**direction_type**", "**assay_type**", "**phenotype**". \
Not required, but highly recommended fields: \
"**platform**", "**sample_type**", "**description**", \
"**covariates**", "**score_cutoff**", "**adj_p_cutoff**", "**logfc_cutoff**". \

"phenotype" is a one or two-word description of a topic, such as a drug treatment, a gene knock out, or a research area. For example, longevity and breast cancer. \
Providing a detailed "description" is highly recommended. For instance, it may include information about how the treatment was administered and how each group was defined. \

Option 1: Create `metadata` by hand. This is not recommended, because typos can occur. \
```
metadata <- list(
  "signature_name" = Myc_reduce_mice_liver_24m,
  "organism" = "Mus Musculus",
  "sample_type" = "liver",
  "phenotype" = "Myc_reduce",
  "direction_type" = "bi-directional",
  "assay_type" = "transcriptomics", 
  "platform" = "GPL6246"
)
```

Option 2: Use `createMetadata()` (recommended). \
This function helps remind you of the built-in attributes. The full list of current built-in attributes is shown [here](https://montilab.github.io/OmicSignature/reference/createMetadata.html). \
You can also provide your own customized attributes using the "others" field. \
```{r}
metadata <- createMetadata(
  # required attributes:
  signature_name = "Myc_reduce_mice_liver_24m",
  organism = "Mus Musculus",
  direction_type = "bi-directional",
  assay_type = "transcriptomics",
  phenotype = "Myc_reduce",

  # optional and recommended:
  covariates = "none",
  description = "mice MYC reduced expression",
  platform = "GPL6246", # use GEO platform ID
  sample_type = "liver", # use BRENDA ontology

  # optional cut-off attributes.
  # specifying them can facilitate the extraction of signatures.
  logfc_cutoff = NULL,
  p_value_cutoff = NULL,
  adj_p_cutoff = 0.05,
  score_cutoff = 5,

  # other optional built-in attributes:
  keywords = c("Myc", "KO", "longevity"),
  cutoff_description = NULL,
  author = NULL,
  PMID = 25619689,
  year = 2015,

  # example of customized attributes:
  others = list("animal_strain" = "C57BL/6")
)
```

#### 2.1.1 "sample_type" and "platform" 
"sample_type" should be a BRENDA ontology term, and "platform" should be a GEO platform accession ID. \
See how to search for the correct term to use in ["BRENDA ontology & GEO platform ID"](https://montilab.github.io/OmicSignature/articles/BRENDA.html). \

#### 2.1.2 "direction_type"
`direction_type` must be one of the following: \
 
- "uni-directional". Only a list of significant feature names is available. Examples including "genes mutated in a disease" and "markers of a cell type". \

- "bi-directional". Significant features can be grouped into "up" and "down" categories. For example, when comparing treatment _vs._ control groups, some features will be higher ("up", or "+") and some will be lower ("down" or "-") in the treatment group. When the phenotype is a continuous trait, such as age, some features will increase ("up", or "+") with age, while others will decrease ("down", or "-") with age. \

- "categorical". Used with multi-valued categorical phenotypes (e.g., "A" _vs._ "B" _vs._ "C"), usually analyzed by ANOVA. \

#### 2.1.3 "assay_type"
`assay_type` is one of the following: \
- "transcriptomics" (e.g. RNA-seq, micro-array) \
- "proteomics" \
- "metabolomics" \
- "methylomics" \
- "genetic_variations" (e.g. SNP, GWAS) \
- "DNA_binding_sites" (e.g. ChIP-seq) \
- "others" \

### 2.2. Differential expression analysis results (difexp)
A differential expression dataframe is **optional** but **highly recommended** if available. It facilitates downstream signature extraction. \

`difexp` is a dataframe with the following **required** columns: \
"**id**", "**symbol**", along with at least one of the following: "**p_value**", "**q_value**", or "**adj_p**". \

"**id**" is a unique identifier for each feature. If not provided, it will be automatically generated. Common examples include unique numbers and probe IDs. \
"**symbol**" is a self-explanatory name for each feature. Common examples include ENSEMBL IDs, UniProt IDs, gene names, and gene symbols. \
"**p_value**", "**q_value**", or "**adj_p**" is the p- or q-value that represents the significance of each feature. \
"score" is a numeric value that represents the importance or significance of a feature. Common examples include t-test statistics and Z-scores. \
Other commonly used optional column examples: "logfc", "beta", "mean", "se", "robust_se", "HR", "annotation". \

Here we use an example out put from the differential expression analysis using the `limma` package. 

```{r}
difexp <- read.table(file.path(system.file("extdata", package = "OmicSignature"), "difmatrix_Myc_mice_liver_24m.txt"),
  header = TRUE, sep = "\t", stringsAsFactors = FALSE
)
head(difexp)
```

We can manually modify the column names to match the requirements. Alternatively, we can use the built-in function `replaceDifexpCol()` designed to replace some 
frequently used alternative column names. 
```{r}
colnames(difexp) <- replaceDifexpCol(colnames(difexp))
head(difexp)
```

### 2.3. Signature 
**signature** is a dataframe with column **"id"** and **"symbol"**. If the signature is "bi-directional" or "categorical" (specified in `direction_type` in `metadata`), then column **"direction"** is also required. \
An optional column "score" is recommended when applicable. \

Option 1: Extract signature from difexp. \
In this example, we create a bi-directional signature from the difexp using the `score_cutoff` and `adj_p_cutoff` specified in the metadata, and use the gene symbol as "symbol". \

```{r}
signature <- difexp %>%
  dplyr::filter(abs(score) > metadata$score_cutoff & adj_p < metadata$adj_p_cutoff) %>%
  dplyr::select(id, symbol, score) %>%
  dplyr::mutate(direction = ifelse(score > 0, "+", "-")) %>%
  dplyr::rename(symbol = "symbol", score = "score")
head(signature)
```

Function `standardizeSigDF()` can help remove duplicated and empty symbols. 
```{r}
signature <- standardizeSigDF(signature)
head(signature)
```

Option 2: Manually create signature dataframe. \
For uni-directional signature: \
```
signature <- data.frame("id" = c(1, 2, 3), "symbol" = c("gene1", "gene2", "gene3"))
```

For bi-directional signature: \
```
signature <- data.frame(
  "id" = c(1, 2, 3),
  "symbol" = c("gene1", "gene2", "gene3"),
  "score" = c(0.45, -3.21, 2.44),
  "direction" = c("+", "-", "+")
)
```

For categorical signature: \
```
signature <- data.frame(
  "id" = c(1, 2, 3, 4),
  "symbol" = c("gene1", "gene2", "gene3", "gene4"),
  "score" = c(0.45, -3.21, 2.44, -2.45),
  "direction" = c("group1", "group1", "group2", "group3")
)
```

### 2.4. Create the `OmicSignature` object
```{r}
OmicObj <- OmicSignature$new(
  metadata = metadata,
  signature = signature,
  difexp = difexp
)
```

Set `print_message` = `TRUE` to see the messages. \
```{r}
OmicObj <- OmicSignature$new(
  metadata = metadata,
  signature = signature,
  difexp = difexp,
  print_message = TRUE
)
```

See the created object information: 
```{r}
print(OmicObj)
```

Use new criteria to extract significant features: \
(this does _not_ change the `signature` saved in the object)
```{r}
OmicObj$extractSignature("abs(score) > 10; adj_p < 0.01")
```

Export the object as a json text file: \
```
writeJson(OmicObj, "Myc_reduce_mice_liver_24m_obj.json")
```

See more in ["Functionalities of OmicSignature"](https://montilab.github.io/OmicSignature/articles/FunOmS.html) section. \

## 3. Create an `OmicSignature` from `difexp` and `metadata`
You can by-pass the generating signature process once you are an expert. Simply provide cutoffs (e.g. `adj_p_cutoff` and `score_cutoff`) in the `metadata`, make sure `difexp` has "adj_p" and "score" columns, and use `OmicSigFromDifexp()` to automatically extract significant features and create the `OmicSignature` object. \

```{r}
OmicObj1 <- OmicSigFromDifexp(difexp, metadata)
OmicObj1
```

</font>
--
