<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Create OmicSignature • OmicSignature</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Create OmicSignature">
<meta property="og:description" content="OmicSignature">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">OmicSignature</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ObjectStructure.html">Structure of OmicSignature &amp; OmicSignatureCollection</a>
    </li>
    <li class="dropdown-header">---- OmicSignature object ----</li>
    <li>
      <a href="../articles/CreateOmS.html">Create OmicSignature</a>
    </li>
    <li>
      <a href="../articles/FunOmS.html">Functionalities, imports and exports</a>
    </li>
    <li>
      <a href="../articles/BRENDA.html">BRENDA ontology &amp; GEO platform ID</a>
    </li>
    <li class="dropdown-header">---- OmicSignatureCollection object ----</li>
    <li>
      <a href="../articles/CreateOmSC.html">Create OmicSignatureCollection</a>
    </li>
    <li>
      <a href="../articles/FunOmSC.html">Functionalities of OmicSignatureCollection</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Create OmicSignature</h1>
                        <h4 data-toc-skip class="author">Vanessa Mengze
Li</h4>
            
            <h4 data-toc-skip class="date">02/27/2024</h4>
      
      
      <div class="hidden name"><code>CreateOmS.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/load_all.html" class="external-link">load_all</a></span><span class="op">(</span><span class="st">"."</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<p><font color="#000000"></font></p>
<div class="section level2">
<h2 id="components-to-create-an-omicsignature-object">1. Components to create an OmicSignature Object<a class="anchor" aria-label="anchor" href="#components-to-create-an-omicsignature-object"></a>
</h2>
<p>An <code>OmicSignature</code> object contains three parts:<br></p>
<ul>
<li><p><strong>metadata</strong>, a list containing metadata
fields.<br>
required fields: “<strong>signature_name</strong>”,
“<strong>organism</strong>”, “<strong>platform</strong>”,
“<strong>direction_type</strong>”, “<strong>assay_type</strong>”,
“<strong>phenotype</strong>”.<br></p></li>
<li><p><strong>signature</strong>, a dataframe.<br>
required columns: “<strong>symbol</strong>”,
“<strong>direction</strong>”<br>
optional column: “<strong>score</strong>”<br></p></li>
<li><p><strong>difexp</strong> (optional), a dataframe of differential
expression analysis results.<br>
required columns: “<strong>id</strong>”, “<strong>symbol</strong>”,
“<strong>score</strong>”, “<strong>p_value</strong>”.<br></p></li>
</ul>
<p>Create the object:<br></p>
<pre><code><span><span class="va">OmicObj</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/OmicSignature.html">OmicSignature</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>  signature <span class="op">=</span> <span class="va">signature</span>,</span>
<span>  difexp <span class="op">=</span> <span class="va">difexp</span></span>
<span><span class="op">)</span></span></code></pre>
<p>You can also read an <code>OmicSignature</code> saved in .json format
using <code><a href="../reference/readJson.html">readJson()</a></code>. See <a href="https://montilab.github.io/OmicSignature/articles/FunOmS.html">“here”</a>
for more details.<br></p>
</div>
<div class="section level2">
<h2 id="create-an-omicsignature-object-step-by-step">2. Create an <code>OmicSignature</code> Object Step-by-Step<a class="anchor" aria-label="anchor" href="#create-an-omicsignature-object-step-by-step"></a>
</h2>
<p>The example provided below is from an experiment for Myc gene reduce
in mice. Signatures were extracted by comparing the liver of treatment
and control when mice is 24-month old. To save space, only the result of
1000 genes are included.<br></p>
<p>This is a bi-directional signature example, which contains up and
down regulated features (genes).<br></p>
<div class="section level3">
<h3 id="metadata">2.1. Metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p><strong>metadata</strong> is a list with the following required
fields:<br>
“<strong>signature_name</strong>”, “<strong>organism</strong>”,
“<strong>direction_type</strong>”, “<strong>assay_type</strong>”.<br>
Not required, but highly recommended fields:<br>
“<strong>phenotype</strong>”, “<strong>platform</strong>”,
“<strong>sample_type</strong>”, “<strong>covariates</strong>”,
“<strong>score_cutoff</strong>”, “<strong>adj_p_cutoff</strong>”,
“<strong>logfc_cutoff</strong>”.<br></p>
<p>Option 1: Create <code>metadata</code> by hand. This is not
recommended, because typos can occur.<br></p>
<pre><code><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"signature_name"</span> <span class="op">=</span> <span class="va">Myc_reduce_mice_liver_24m</span>,</span>
<span>  <span class="st">"organism"</span> <span class="op">=</span> <span class="st">"Mus Musculus"</span>,</span>
<span>  <span class="st">"sample_type"</span> <span class="op">=</span> <span class="st">"liver"</span>,</span>
<span>  <span class="st">"phenotype"</span> <span class="op">=</span> <span class="st">"Myc_reduce"</span>,</span>
<span>  <span class="st">"direction_type"</span> <span class="op">=</span> <span class="st">"bi-directional"</span>,</span>
<span>  <span class="st">"assay_type"</span> <span class="op">=</span> <span class="st">"transcriptomics"</span>, </span>
<span>  <span class="st">"platform"</span> <span class="op">=</span> <span class="st">"GPL6246"</span>,</span>
<span>  <span class="st">"adj_p_cutoff"</span> <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  <span class="st">"score_cutoff"</span> <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  <span class="st">"keywords"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myc"</span>, <span class="st">"KO"</span>, <span class="st">"longevity"</span><span class="op">)</span>,</span>
<span>  <span class="st">"PMID"</span> <span class="op">=</span> <span class="fl">25619689</span>,</span>
<span>  <span class="st">"year"</span> <span class="op">=</span> <span class="fl">2015</span></span>
<span><span class="op">)</span></span></code></pre>
<p>Option 2: Use function <code><a href="../reference/createMetadata.html">createMetadata()</a></code>
(recommended).<br>
This function helps remind you of the built-in attributes. The full list
of current built-in attributes is shown <a href="https://montilab.github.io/OmicSignature/reference/createMetadata.html">here</a>.<br>
You can also provide your own customized attributes.<br></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createMetadata.html">createMetadata</a></span><span class="op">(</span></span>
<span>  <span class="co"># required attributes:</span></span>
<span>  signature_name <span class="op">=</span> <span class="st">"Myc_reduce_mice_liver_24m"</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Mus Musculus"</span>,</span>
<span>  direction_type <span class="op">=</span> <span class="st">"bi-directional"</span>,</span>
<span>  assay_type <span class="op">=</span> <span class="st">"transcriptomics"</span>,</span>
<span></span>
<span>  <span class="co"># optional and recommended:</span></span>
<span>  phenotype <span class="op">=</span> <span class="st">"Myc_reduce"</span>,</span>
<span>  covariates <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  platform <span class="op">=</span> <span class="st">"GPL6246"</span>, <span class="co"># must be a GEO platform ID</span></span>
<span>  sample_type <span class="op">=</span> <span class="st">"liver"</span>, <span class="co"># must be BRENDA ontology</span></span>
<span></span>
<span>  <span class="co"># optional cut-off attributes.</span></span>
<span>  <span class="co"># specifying them can facilitate the extraction of signatures.</span></span>
<span>  logfc_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p_value_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  adj_p_cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  score_cutoff <span class="op">=</span> <span class="fl">5</span>,</span>
<span></span>
<span>  <span class="co"># other optional built-in attributes:</span></span>
<span>  keywords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myc"</span>, <span class="st">"KO"</span>, <span class="st">"longevity"</span><span class="op">)</span>,</span>
<span>  cutoff_description <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  author <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  PMID <span class="op">=</span> <span class="fl">25619689</span>,</span>
<span>  year <span class="op">=</span> <span class="fl">2015</span>,</span>
<span></span>
<span>  <span class="co"># example of a customized attribute:</span></span>
<span>  animal_strain <span class="op">=</span> <span class="st">"C57BL/6"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="instruction-for-sample_type-and-platform">2.1.1 Instruction for “sample_type” and “platform”<a class="anchor" aria-label="anchor" href="#instruction-for-sample_type-and-platform"></a>
</h4>
<p>“sample_type” should be a BRENDA ontology term, and “platform” should
be a GEO platform accession ID. Otherwise, you will get warnings. See
how to search for the correct term to use in <a href="https://montilab.github.io/OmicSignature/articles/BRENDA.html">“BRENDA
ontology &amp; GEO platform ID”</a>.<br></p>
</div>
<div class="section level4">
<h4 id="instruction-for-direction_type">2.1.2 Instruction for “direction_type”<a class="anchor" aria-label="anchor" href="#instruction-for-direction_type"></a>
</h4>
<p><code>direction_type</code> must be one of the following:<br></p>
<ul>
<li><p>“uni-directional”. You only have a list of significant feature
names, and don’t know if they are up or down regulated in the treatment
group, or directional information is not applicable. An example would be
“genes mutated in a disease.”<br></p></li>
<li><p>“bi-directional”. Significant features can be grouped into “up”
and “down” categories. For example, when comparing treatment
<em>vs.</em> control groups, some features will be higher (“up”, or “+”)
and some will be lower (“down” or “-”) in the treatment group.
Similarly, when the phenotype is a continuous trait, such as age, some
features will increase (“up”, or “+”) with age, while others will
decrease (“down”, or “-”).<br></p></li>
<li><p>“multiple”. Used with multi-valued categorical phenotypes (e.g.,
“low” <em>vs.</em> “medium” <em>vs.</em> “high”), usually analyzed by
ANOVA. In this case, the “direction” column in signature table should be
the phenotype’s category (e.g., “low”).<br></p></li>
</ul>
</div>
<div class="section level4">
<h4 id="instruction-for-assay_type">2.1.3 Instruction for “assay_type”<a class="anchor" aria-label="anchor" href="#instruction-for-assay_type"></a>
</h4>
<p>For the time being, <code>assay_type</code> is one of the
following:<br>
- “transcriptomics”<br>
- “proteomics”<br>
- “metabolomics”<br>
- “methylomics”<br>
- “genetic_variations”<br>
- “others”<br></p>
</div>
</div>
<div class="section level3">
<h3 id="diffanal-results-difexp">2.2. Diffanal results (difexp)<a class="anchor" aria-label="anchor" href="#diffanal-results-difexp"></a>
</h3>
<p>A differential expression dataframe is <strong>optional</strong> but
<strong>highly recommended</strong> if available. It facilitates
downstream signature extraction.<br></p>
<p><code>difexp</code> is a dataframe with <strong>required</strong>
columns:<br>
“<strong>id</strong>”, “<strong>symbol</strong>”,
“<strong>score</strong>”, “<strong>p_value</strong>”.<br>
“id” is used as an unique identifier in case there are duplicated
symbols. Frequently used examples including: probe ID, ENSEMBL ID,
UniProt ID, unique numbers.<br>
“score” is usually t-test statistics or Z-score.<br></p>
<p>Some frequently used optional column names:<br>
“logfc”, “est”, “aveexpr”, “se”, “robust_se”, “HR”, “adj_p”,
“gene_name”, “gene_annotation”.<br></p>
<p>Here we use an example out put from the differential expression
analysis using the <code>limma</code> package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">difexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"OmicSignature"</span><span class="op">)</span>, <span class="st">"difmatrix_Myc_mice_liver_24m.txt"</span><span class="op">)</span>,</span>
<span>  header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span></span>
<span><span class="co">#&gt;    logFC AveExpr      t P.Value adj.P.Val      b Probe.ID gene_symbol</span></span>
<span><span class="co">#&gt; 1 -0.093   6.677 -1.030   0.338     0.702 -6.371 10344614     Gm16088</span></span>
<span><span class="co">#&gt; 2 -0.098   2.998 -1.116   0.302     0.674 -6.281 10344616     Gm26206</span></span>
<span><span class="co">#&gt; 3 -0.215   5.055 -2.589   0.037     0.287 -4.287 10344620     Gm10568</span></span>
<span><span class="co">#&gt; 4 -0.122  12.057 -1.588   0.157     0.526 -5.714 10344624      Lypla1</span></span>
<span><span class="co">#&gt; 5  0.014  10.087  0.222   0.831     0.953 -6.903 10344633       Tcea1</span></span>
<span><span class="co">#&gt; 6 -0.034   9.947 -0.464   0.657     0.884 -6.810 10344637     Atp6v1h</span></span></code></pre></div>
<p>We can manually modify the column names to match the requirements.
Alternatively, we can use the built-in function
<code><a href="../reference/replaceDifexpCol.html">replaceDifexpCol()</a></code> designed to replace some frequently used
alternative column names.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/replaceDifexpCol.html">replaceDifexpCol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span></span>
<span><span class="co">#&gt;    logfc aveexpr  score p_value adj_p      b       id  symbol</span></span>
<span><span class="co">#&gt; 1 -0.093   6.677 -1.030   0.338 0.702 -6.371 10344614 Gm16088</span></span>
<span><span class="co">#&gt; 2 -0.098   2.998 -1.116   0.302 0.674 -6.281 10344616 Gm26206</span></span>
<span><span class="co">#&gt; 3 -0.215   5.055 -2.589   0.037 0.287 -4.287 10344620 Gm10568</span></span>
<span><span class="co">#&gt; 4 -0.122  12.057 -1.588   0.157 0.526 -5.714 10344624  Lypla1</span></span>
<span><span class="co">#&gt; 5  0.014  10.087  0.222   0.831 0.953 -6.903 10344633   Tcea1</span></span>
<span><span class="co">#&gt; 6 -0.034   9.947 -0.464   0.657 0.884 -6.810 10344637 Atp6v1h</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="signature">2.3. Signature<a class="anchor" aria-label="anchor" href="#signature"></a>
</h3>
<p><strong>signature</strong> is a dataframe with column
<strong>“symbol”</strong>. “symbol” should be a subset of “symbol” in
difexp if present.<br>
Also, if the signature is “bi-directional” or “multiple” (specified in
<code>direction_type</code> in <code>metadata</code> list), then column
<strong>“direction”</strong> is also required.<br>
An optional column “score” is recommended when feature scores are
available.<br></p>
<p>Option 1: Extract signature from difexp.<br>
Here we create a bi-directional signature manually from the difexp
generated above, using the <code>score_cutoff</code> and
<code>adj_p_cutoff</code> previously specified in the metadata.<br></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">signature</span> <span class="op">&lt;-</span> <span class="va">difexp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">metadata</span><span class="op">$</span><span class="va">score_cutoff</span> <span class="op">&amp;</span> <span class="va">adj_p</span> <span class="op">&lt;</span> <span class="va">metadata</span><span class="op">$</span><span class="va">adj_p_cutoff</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">symbol</span>, <span class="va">score</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>direction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">score</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="st">"symbol"</span>, score <span class="op">=</span> <span class="st">"score"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span></span>
<span><span class="co">#&gt;     symbol   score direction</span></span>
<span><span class="co">#&gt; 1 n-R5s210  -5.953         -</span></span>
<span><span class="co">#&gt; 2    Il1r1 -13.543         -</span></span>
<span><span class="co">#&gt; 3   Adam23  -6.990         -</span></span>
<span><span class="co">#&gt; 4   Mogat1   5.496         +</span></span>
<span><span class="co">#&gt; 5   Utp14b  -5.404         -</span></span>
<span><span class="co">#&gt; 6     Ctse  14.762         +</span></span></code></pre></div>
<p>(note: if you see numbers instead of gene symbol name in the first
column, please check if the “symbol” column in your difexp matrix is
“character” and not accidentally be “factor”)<br></p>
<p>Function <code><a href="../reference/standardizeSigDF.html">standardizeSigDF()</a></code> can help remove duplicate
rows, empty symbols in the signature dataframe.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/standardizeSigDF.html">standardizeSigDF</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span></span>
<span><span class="co">#&gt;    symbol   score direction</span></span>
<span><span class="co">#&gt; 1   Chil1 -25.413         -</span></span>
<span><span class="co">#&gt; 2    Ctse  14.762         +</span></span>
<span><span class="co">#&gt; 3   Il1r1 -13.543         -</span></span>
<span><span class="co">#&gt; 4    Eya1  10.487         +</span></span>
<span><span class="co">#&gt; 5    Myl1 -10.315         -</span></span>
<span><span class="co">#&gt; 6 Sh2d1b1   8.818         +</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span></span>
<span><span class="co">#&gt;      symbol  score direction</span></span>
<span><span class="co">#&gt; 13    Erbb4  6.083         +</span></span>
<span><span class="co">#&gt; 14 n-R5s210 -5.953         -</span></span>
<span><span class="co">#&gt; 15   Lpgat1 -5.574         -</span></span>
<span><span class="co">#&gt; 16   Mogat1  5.496         +</span></span>
<span><span class="co">#&gt; 17    Ildr2 -5.429         -</span></span>
<span><span class="co">#&gt; 18   Utp14b -5.404         -</span></span></code></pre></div>
<p>Option 2: Manually write signature.<br>
For uni-directional signature:<br></p>
<pre><code><span><span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span></span></code></pre>
<p>For bi-directional signature:<br></p>
<pre><code><span><span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.45</span>, <span class="op">-</span><span class="fl">3.21</span>, <span class="fl">2.44</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span></span></code></pre>
<p>The direction will be automatically determined by the score value
provided.<br></p>
</div>
<div class="section level3">
<h3 id="create-the-omicsignature-object">2.4. Create the <code>OmicSignature</code> object<a class="anchor" aria-label="anchor" href="#create-the-omicsignature-object"></a>
</h3>
<p>Use <code>OmicSignature$new()</code> to create a new OmicSignature R6
object.<br></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OmicObj</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/OmicSignature.html">OmicSignature</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>  signature <span class="op">=</span> <span class="va">signature</span>,</span>
<span>  difexp <span class="op">=</span> <span class="va">difexp</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;   [Success] OmicSignature object Myc_reduce_mice_liver_24m created.</span></span></code></pre></div>
<p>You can also ask the program to print the messages while creating the
OmicSignature Object. By default, <code>print_message</code> is set to
be <code>FALSE</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OmicObj</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/OmicSignature.html">OmicSignature</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>  signature <span class="op">=</span> <span class="va">signature</span>,</span>
<span>  difexp <span class="op">=</span> <span class="va">difexp</span>,</span>
<span>  print_message <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;   --Required attributes for metadata: signature_name, organism, direction_type, assay_type --</span></span>
<span><span class="co">#&gt;   [Success] Metadata is saved. </span></span>
<span><span class="co">#&gt;   [Success] Signature is valid. </span></span>
<span><span class="co">#&gt;   difexp: additional columns found: logfc, aveexpr, b. </span></span>
<span><span class="co">#&gt;   [Success] difexp matrix is valid. </span></span>
<span><span class="co">#&gt;   [Success] OmicSignature object Myc_reduce_mice_liver_24m created.</span></span></code></pre></div>
<p>Use <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> to see information:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">OmicObj</span><span class="op">)</span></span>
<span><span class="co">#&gt; Signature Object: </span></span>
<span><span class="co">#&gt;   Metadata: </span></span>
<span><span class="co">#&gt;     adj_p_cutoff = 0.05 </span></span>
<span><span class="co">#&gt;     animal_strain = C57BL/6 </span></span>
<span><span class="co">#&gt;     assay_type = transcriptomics </span></span>
<span><span class="co">#&gt;     covariates = none </span></span>
<span><span class="co">#&gt;     direction_type = bi-directional </span></span>
<span><span class="co">#&gt;     keywords = Myc, KO, longevity </span></span>
<span><span class="co">#&gt;     organism = Mus Musculus </span></span>
<span><span class="co">#&gt;     phenotype = Myc_reduce </span></span>
<span><span class="co">#&gt;     platform = GPL6246 </span></span>
<span><span class="co">#&gt;     PMID = 25619689 </span></span>
<span><span class="co">#&gt;     sample_type = liver </span></span>
<span><span class="co">#&gt;     score_cutoff = 5 </span></span>
<span><span class="co">#&gt;     signature_name = Myc_reduce_mice_liver_24m </span></span>
<span><span class="co">#&gt;     year = 2015 </span></span>
<span><span class="co">#&gt;   signature: </span></span>
<span><span class="co">#&gt;     - (12)</span></span>
<span><span class="co">#&gt;     + (6)</span></span>
<span><span class="co">#&gt;   Differential Expression Data: </span></span>
<span><span class="co">#&gt;     1000 x 8</span></span></code></pre></div>
<p>Use new criteria to extract significant features:<br>
(note: this does <em>not</em> change the <code>signature</code> saved in
the object)</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OmicObj</span><span class="op">$</span><span class="fu">extractSignature</span><span class="op">(</span><span class="st">"abs(score) &gt; 10; adj_p &lt; 0.01"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   symbol   score direction</span></span>
<span><span class="co">#&gt; 1  Chil1 -25.413         -</span></span>
<span><span class="co">#&gt; 2   Ctse  14.762         +</span></span>
<span><span class="co">#&gt; 3  Il1r1 -13.543         -</span></span>
<span><span class="co">#&gt; 4   Eya1  10.487         +</span></span>
<span><span class="co">#&gt; 5   Myl1 -10.315         -</span></span></code></pre></div>
<p>Write it in a json text format:</p>
<pre><code><span><span class="fu"><a href="../reference/writeJson.html">writeJson</a></span><span class="op">(</span><span class="va">OmicObj</span>, <span class="st">"Myc_reduce_mice_liver_24m_obj.json"</span><span class="op">)</span></span></code></pre>
<p>See more in <a href="https://montilab.github.io/OmicSignature/articles/FunOmS.html">“Functionalities
of OmicSignature”</a> section.<br></p>
</div>
</div>
<div class="section level2">
<h2 id="create-an-omicsignature-from-difexp-and-metadata">3. Create an <code>OmicSignature</code> from <code>difexp</code> and
<code>metadata</code><a class="anchor" aria-label="anchor" href="#create-an-omicsignature-from-difexp-and-metadata"></a>
</h2>
<p>You can by-pass the generating signature process once you are an
expert. Simply provide cutoffs (e.g. <code>adj_p_cutoff</code> and
<code>score_cutoff</code>) in the <code>metadata</code>, make sure
<code>difexp</code> has those columns available, and use
<code><a href="../reference/OmicSigFromDifexp.html">OmicSigFromDifexp()</a></code> to extract significant features and
create the <code>OmicSignature</code> object.<br></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OmicObj1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/OmicSigFromDifexp.html">OmicSigFromDifexp</a></span><span class="op">(</span><span class="va">difexp</span>, <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt; -- criterias used to extract signatures:  abs(score) &gt; 5; adj_p &lt; 0.05 . </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   [Success] OmicSignature object Myc_reduce_mice_liver_24m created.</span></span>
<span><span class="va">OmicObj1</span></span>
<span><span class="co">#&gt; Signature Object: </span></span>
<span><span class="co">#&gt;   Metadata: </span></span>
<span><span class="co">#&gt;     adj_p_cutoff = 0.05 </span></span>
<span><span class="co">#&gt;     animal_strain = C57BL/6 </span></span>
<span><span class="co">#&gt;     assay_type = transcriptomics </span></span>
<span><span class="co">#&gt;     covariates = none </span></span>
<span><span class="co">#&gt;     direction_type = bi-directional </span></span>
<span><span class="co">#&gt;     keywords = Myc, KO, longevity </span></span>
<span><span class="co">#&gt;     organism = Mus Musculus </span></span>
<span><span class="co">#&gt;     phenotype = Myc_reduce </span></span>
<span><span class="co">#&gt;     platform = GPL6246 </span></span>
<span><span class="co">#&gt;     PMID = 25619689 </span></span>
<span><span class="co">#&gt;     sample_type = liver </span></span>
<span><span class="co">#&gt;     score_cutoff = 5 </span></span>
<span><span class="co">#&gt;     signature_name = Myc_reduce_mice_liver_24m </span></span>
<span><span class="co">#&gt;     year = 2015 </span></span>
<span><span class="co">#&gt;   signature: </span></span>
<span><span class="co">#&gt;     - (12)</span></span>
<span><span class="co">#&gt;     + (6)</span></span>
<span><span class="co">#&gt;   Differential Expression Data: </span></span>
<span><span class="co">#&gt;     1000 x 8</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="section"></h2>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Mengze Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
