<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>OmicSignature • OmicSignature</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="OmicSignature">
<meta property="og:description" content="OmicSignature">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">OmicSignature</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/BRENDA.html">BRENDA</a>
    </li>
    <li>
      <a href="../articles/OmicSigCol_vig.html">OmicSignatureCollection</a>
    </li>
    <li>
      <a href="../articles/OmicSig_vig.html">OmicSignature</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="OmicSig_vig_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>OmicSignature</h1>
                        <h4 class="author">Vanessa Mengze Li</h4>
            
            <h4 class="date">Nov 22nd 2020</h4>
      
      
      <div class="hidden name"><code>OmicSig_vig.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/load_all.html">load_all</a></span><span class="op">(</span><span class="st">"."</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></code></pre></div>
<div id="create-an-omicsignature-object" class="section level2">
<h2 class="hasAnchor">
<a href="#create-an-omicsignature-object" class="anchor"></a>Create an OmicSignature Object</h2>
<p>An <code>OmicSignature</code> object contains three parts:<br>
- <strong>metadata</strong>, a list contains metadata fields;<br>
- <strong>signature</strong>, a dataframe with feature symbols and directions, and scores if applicable;<br>
- <strong>difexp</strong>, optional, a dataframe from differential expression analysis result.</p>
<p>The example provided below is from an experiment for Myc gene reduce in mice. Signatures was extracted by comparing the liver of treatment and control when mice is 24-month old. This is a bi-directional signature example, which contains up and down regulated features (genes).</p>
<div id="specify-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#specify-metadata" class="anchor"></a>Specify metadata</h3>
<p>To be saved into a OmicSignature object, the required <strong>metadata</strong> fields are:<br><strong>“signature_name”, “organism”, “platform”, “direction_type”, “phenotype”</strong>.<br>
We can also specify other criterias such as adj_p (fdr corrected p-value) and log fold change cutoff of the signature, and provide key words of our experiment to make the description more complete.<br>
One option is to create metadata list by hand, similar as follows:</p>
<pre><code>metadata &lt;- list(
  "signature_name" = Myc_reduce_mice_liver_24m,
  "organism" = "Mus Musculus",
  "sample_type" = "liver",
  "phenotype" = "Myc_reduce",
  "direction_type" = "bi-directional",
  "platform" = "GPL6246",
  "fdr_cutoff" = 0.05,
  "score_cutoff" = 7,
  "keywords" = c("Myc", "KO", "longevity"),
  "PMID" = 25619689,
  "year" = 2015
)</code></pre>
<p>Or use our build-in function <code>createMetadata</code> (recommended):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createMetadata.html">createMetadata</a></span><span class="op">(</span>
  signature_name <span class="op">=</span> <span class="st">"Myc_reduce_mice_liver_24m"</span>,
  organism <span class="op">=</span> <span class="st">"Mus Musculus"</span>,
  sample_type <span class="op">=</span> <span class="st">"liver"</span>,
  phenotype <span class="op">=</span> <span class="st">"Myc_reduce"</span>,
  direction_type <span class="op">=</span> <span class="st">"bi-directional"</span>,
  platform <span class="op">=</span> <span class="st">"GPL6246"</span>,
  fdr_cutoff <span class="op">=</span> <span class="fl">0.05</span>,
  score_cutoff <span class="op">=</span> <span class="fl">7</span>,
  keywords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Myc"</span>, <span class="st">"KO"</span>, <span class="st">"longevity"</span><span class="op">)</span>,
  PMID <span class="op">=</span> <span class="fl">25619689</span>,
  year <span class="op">=</span> <span class="fl">2015</span>
<span class="op">)</span></code></pre></div>
<p>If “sample_type” is not a BRENDA ontology name or “platform” is not a valid GEO platform accession ID, you will receive warnings. See more info below.</p>
<div id="additional-info-for-sample_type-and-platform" class="section level4">
<h4 class="hasAnchor">
<a href="#additional-info-for-sample_type-and-platform" class="anchor"></a>Additional info for “sample_type” and “platform”</h4>
<p>If your signature was generated from a cell line or tissue sample, we highly recommended you to specify the <strong>cell or tissue name using the standard BRENDA ontology term</strong>. e.g. “SUM-149PT cell”, and not “SUM 149 PT”.<br>
You can find the correct ontology term by using function <code><a href="../reference/BRENDASearch.html">BRENDASearch()</a></code>.<br>
For example, search for ontology terms contain “SUM” and “cell”:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/BRENDASearch.html">BRENDASearch</a></span><span class="op">(</span><span class="st">"SUM cell"</span><span class="op">)</span>
<span class="co">#&gt;             ID            Name              </span>
<span class="co">#&gt; BTO:0004928 "BTO:0004928" "SUM-159PT cell"  </span>
<span class="co">#&gt; BTO:0005387 "BTO:0005387" "SUM-102PT cell"  </span>
<span class="co">#&gt; BTO:0005389 "BTO:0005389" "SUM-44PE cell"   </span>
<span class="co">#&gt; BTO:0005390 "BTO:0005390" "SUM-52PE cell"   </span>
<span class="co">#&gt; BTO:0005475 "BTO:0005475" "SUM-149PT cell"  </span>
<span class="co">#&gt; BTO:0005476 "BTO:0005476" "SUM-185PE cell"  </span>
<span class="co">#&gt; BTO:0005477 "BTO:0005477" "SUM-190PT cell"  </span>
<span class="co">#&gt; BTO:0005478 "BTO:0005478" "SUM-225CWN cell" </span>
<span class="co">#&gt; BTO:0005479 "BTO:0005479" "SUM-229PE cell"  </span>
<span class="co">#&gt; BTO:0005932 "BTO:0005932" "SUM cell"        </span>
<span class="co">#&gt; BTO:0005933 "BTO:0005933" "SUM-1315MO2 cell"</span></code></pre></div>
<p>In addition, we use GEO platform accession ID, e.g. “GPL6246”, in our <em>SigRepo database</em>. If your data was downloaded from GEO, you can usually find it in the website. If you are not sure about your GEO platform accession ID, you can search by using function <code><a href="../reference/GEOPlatformSearch.html">GEOPlatformSearch()</a></code>.<br>
For example, search for platform names contains “illumina” and species name contains “homo”:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/GEOPlatformSearch.html">GEOPlatformSearch</a></span><span class="op">(</span><span class="st">"illumina"</span>, species <span class="op">=</span> <span class="st">"homo"</span><span class="op">)</span>
<span class="co">#&gt;      Accession                                       Title</span>
<span class="co">#&gt; 737   GPL27991         illumina Nextseq 500 (Homo sapiens)</span>
<span class="co">#&gt; 3246  GPL24432          illumina HiSeq 2500 (Homo sapiens)</span>
<span class="co">#&gt; 4224  GPL23177         illumina NextSeq 500 (Homo sapiens)</span>
<span class="co">#&gt; 4344  GPL23049 illumina Genome Analyzer IIx (Homo sapiens)</span>
<span class="co">#&gt;                      Technology     Taxonomy Data.Rows Samples.Count</span>
<span class="co">#&gt; 737  high-throughput sequencing Homo sapiens         0             0</span>
<span class="co">#&gt; 3246 high-throughput sequencing Homo sapiens         0             8</span>
<span class="co">#&gt; 4224 high-throughput sequencing Homo sapiens         0            13</span>
<span class="co">#&gt; 4344 high-throughput sequencing Homo sapiens         0             2</span>
<span class="co">#&gt;      Series.Count Contact Release.Date</span>
<span class="co">#&gt; 737             0     GEO Jan 13, 2020</span>
<span class="co">#&gt; 3246            2     GEO Dec 26, 2017</span>
<span class="co">#&gt; 4224            1     GEO Mar 14, 2017</span>
<span class="co">#&gt; 4344            1     GEO Feb 09, 2017</span></code></pre></div>
</div>
<div id="additional-info-for-direction_type" class="section level4">
<h4 class="hasAnchor">
<a href="#additional-info-for-direction_type" class="anchor"></a>Additional info for “direction_type”</h4>
<p><code>direction_type</code> is either:<br>
- “uni-directional”. You only have a list of significant feature names but don’t know if they are up or down regulated in the treatment group, or directional infomation is not applicable.<br>
- “bi-directional”. In most cases significant features can be grouped into “up” and “down” regulated features. For example, when comparing treatment group with control group, some features will be higher (“up”, or “+”) and some will be lower (“down” or “-”) in treatment.<br>
- “multi-directional”. Used when comparing multiple sampels together, usually using ANOVA. In this case, the “direction” column in signature table should be group name.</p>
</div>
<div id="add-more-info-in-metadata" class="section level4">
<h4 class="hasAnchor">
<a href="#add-more-info-in-metadata" class="anchor"></a>Add more info in metadata</h4>
<p>In this example, we are working with mouse strain C57BL/6. We can add this information into our <code>metadata</code>.<br><code>metadata</code> is just a list. Add new terms into this list like you normally do.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metadata</span><span class="op">$</span><span class="st">"animal_strain"</span> <span class="op">&lt;-</span> <span class="st">"C57BL/6"</span></code></pre></div>
</div>
</div>
<div id="differential-expression-analysis-result-difexp" class="section level3">
<h3 class="hasAnchor">
<a href="#differential-expression-analysis-result-difexp" class="anchor"></a>Differential expression analysis result <code>difexp</code>
</h3>
<p>A differential expression analysis matrix is <strong>optional</strong> but <strong>highly recommended</strong> if you have it.<br>
The differential expression analysis in this example was performed using a standard pipeline from lm package. Available columns are logFC, AveExpr, t score, P.Value, adj.P.Val, B score, Probe.ID, gene_symbol and gene_name.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">difexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"OmicSignature"</span><span class="op">)</span>, <span class="st">"difmatrix_Myc_mice_liver_24m_raw.txt"</span><span class="op">)</span>,
  header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span>
<span class="co">#&gt;         logFC   AveExpr          t    P.Value adj.P.Val         b Probe.ID</span>
<span class="co">#&gt; 1 -0.09301000  6.676552 -1.0304955 0.33776819 0.7015182 -6.370549 10344614</span>
<span class="co">#&gt; 2 -0.09837667  2.998012 -1.1161772 0.30194862 0.6735042 -6.281333 10344616</span>
<span class="co">#&gt; 3 -0.21524000  5.055207 -2.5893972 0.03664519 0.2872695 -4.287330 10344620</span>
<span class="co">#&gt; 4 -0.12186667 12.056667 -1.5881487 0.15719140 0.5256233 -5.713531 10344624</span>
<span class="co">#&gt; 5  0.01440000 10.087133  0.2221561 0.83066749 0.9532330 -6.902609 10344633</span>
<span class="co">#&gt; 6 -0.03361667  9.947465 -0.4638379 0.65713620 0.8842391 -6.810300 10344637</span>
<span class="co">#&gt;   gene_symbol                                       gene_name</span>
<span class="co">#&gt; 1     Gm16088                            predicted gene 16088</span>
<span class="co">#&gt; 2     Gm26206                           predicted gene, 26206</span>
<span class="co">#&gt; 3     Gm10568                            predicted gene 10568</span>
<span class="co">#&gt; 4      Lypla1                             lysophospholipase 1</span>
<span class="co">#&gt; 5       Tcea1       transcription elongation factor A (SII) 1</span>
<span class="co">#&gt; 6     Atp6v1h ATPase, H+ transporting, lysosomal V1 subunit H</span></code></pre></div>
<p>To be saved into a OmicSignature object, the required columns for <strong>differential expression matrix</strong> are:<br><strong>“id”, “symbol”, “score”, “adj_p”</strong>.<br>
“id” is used as an unique identifier in case there are duplicated gene symbols. Frequently used id including probe ID, ENSEMBL ID, or unique numbers.</p>
<p>Modify the column names to match the requirement.<br>
Our build-in function <code><a href="../reference/replaceDifexpCol.html">replaceDifexpCol()</a></code> is designed to replace some frequently-used alternative column names. You can also change the column names manully.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/replaceDifexpCol.html">replaceDifexpCol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span>
<span class="co">#&gt;         logfc   aveexpr      score    p_value     adj_p         b       id</span>
<span class="co">#&gt; 1 -0.09301000  6.676552 -1.0304955 0.33776819 0.7015182 -6.370549 10344614</span>
<span class="co">#&gt; 2 -0.09837667  2.998012 -1.1161772 0.30194862 0.6735042 -6.281333 10344616</span>
<span class="co">#&gt; 3 -0.21524000  5.055207 -2.5893972 0.03664519 0.2872695 -4.287330 10344620</span>
<span class="co">#&gt; 4 -0.12186667 12.056667 -1.5881487 0.15719140 0.5256233 -5.713531 10344624</span>
<span class="co">#&gt; 5  0.01440000 10.087133  0.2221561 0.83066749 0.9532330 -6.902609 10344633</span>
<span class="co">#&gt; 6 -0.03361667  9.947465 -0.4638379 0.65713620 0.8842391 -6.810300 10344637</span>
<span class="co">#&gt;    symbol                                       gene_name</span>
<span class="co">#&gt; 1 Gm16088                            predicted gene 16088</span>
<span class="co">#&gt; 2 Gm26206                           predicted gene, 26206</span>
<span class="co">#&gt; 3 Gm10568                            predicted gene 10568</span>
<span class="co">#&gt; 4  Lypla1                             lysophospholipase 1</span>
<span class="co">#&gt; 5   Tcea1       transcription elongation factor A (SII) 1</span>
<span class="co">#&gt; 6 Atp6v1h ATPase, H+ transporting, lysosomal V1 subunit H</span></code></pre></div>
</div>
<div id="create-signature" class="section level3">
<h3 class="hasAnchor">
<a href="#create-signature" class="anchor"></a>Create signature</h3>
<p>This is a bi-directional signature example. It contains up and down regulated genes.<br>
There are multiple ways to get signatures from a differential expression analysis matrix. One way is to use <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> from <code>dplyr</code> package.<br>
In this example, we previously specified the <code>score_cutoff</code> and <code>fdr_cutoff</code> when creating the metadata. We can use them here.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">temp_upsig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">difexp</span>, <span class="va">score</span> <span class="op">&gt;</span> <span class="va">metadata</span><span class="op">$</span><span class="va">score_cutoff</span> <span class="op">&amp;</span> <span class="va">adj_p</span> <span class="op">&lt;</span> <span class="va">metadata</span><span class="op">$</span><span class="va">fdr_cutoff</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>,
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">difexp</span>, <span class="va">score</span> <span class="op">&gt;</span> <span class="va">metadata</span><span class="op">$</span><span class="va">score_cutoff</span> <span class="op">&amp;</span> <span class="va">adj_p</span> <span class="op">&lt;</span> <span class="va">metadata</span><span class="op">$</span><span class="va">fdr_cutoff</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span>,
  <span class="st">"+"</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">temp_upsig</span><span class="op">)</span>
<span class="co">#&gt;      [,1]       [,2]               [,3]</span>
<span class="co">#&gt; [1,] "Ctse"     "14.7620707119707" "+" </span>
<span class="co">#&gt; [2,] "Sh2d1b1"  "8.81828143794327" "+" </span>
<span class="co">#&gt; [3,] "Eya1"     "10.486943196476"  "+" </span>
<span class="co">#&gt; [4,] "Serpine2" "7.82916045128826" "+" </span>
<span class="co">#&gt; [5,] "Fcer1g"   "7.11819488945291" "+" </span>
<span class="co">#&gt; [6,] "Slamf7"   "8.59861880131375" "+"</span>

<span class="va">temp_dnsig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">difexp</span>, <span class="va">score</span> <span class="op">&lt;</span> <span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="va">metadata</span><span class="op">$</span><span class="va">score_cutoff</span> <span class="op">&amp;</span> <span class="va">adj_p</span> <span class="op">&lt;</span> <span class="va">metadata</span><span class="op">$</span><span class="va">fdr_cutoff</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>,
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">difexp</span>, <span class="va">score</span> <span class="op">&lt;</span> <span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="va">metadata</span><span class="op">$</span><span class="va">score_cutoff</span> <span class="op">&amp;</span> <span class="va">adj_p</span> <span class="op">&lt;</span> <span class="va">metadata</span><span class="op">$</span><span class="va">fdr_cutoff</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span>,
  <span class="st">"-"</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">temp_dnsig</span><span class="op">)</span>
<span class="co">#&gt;      [,1]      [,2]                [,3]</span>
<span class="co">#&gt; [1,] "Il1r1"   "-13.5427340844379" "-" </span>
<span class="co">#&gt; [2,] "Chil1"   "-25.4131776388419" "-" </span>
<span class="co">#&gt; [3,] "Kcnt2"   "-7.72798204361239" "-" </span>
<span class="co">#&gt; [4,] "Olfr16"  "-7.01030444724159" "-" </span>
<span class="co">#&gt; [5,] "Ankrd23" "-7.86723140130081" "-" </span>
<span class="co">#&gt; [6,] "Myl1"    "-10.3151891073448" "-"</span></code></pre></div>
<p>(note: if you see numbers instead of gene symbol name in the first column, please check if the “symbol” column in your difexp matrix is “character” and not accidentally be “factor”)</p>
<p>To be stored into OmicSignature object, <strong>signature</strong> need to be a <strong>dataframe</strong> with column <strong>“signature_symbol”</strong>.<br>
If the signature is “bi-directional” or “multi-directional” (specified in <code>direction_type</code> in <code>metadata</code> list), then column <strong>“signature_direction”</strong> is also required. “uni-directional” type does not require this column.<br>
An <strong>optional column “signature_score”</strong> is used when feature scores are available.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">signatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">temp_upsig</span>, <span class="va">temp_dnsig</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">signatures</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"signature_symbol"</span>, <span class="st">"signature_score"</span>, <span class="st">"signature_direction"</span><span class="op">)</span></code></pre></div>
<p>Our function <code><a href="../reference/standardizeSigDF.html">standardizeSigDF()</a></code> can help you to remove duplicate rows, empty symbols in the signature dataframe, if any.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">signatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/standardizeSigDF.html">standardizeSigDF</a></span><span class="op">(</span><span class="va">signatures</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">signatures</span><span class="op">)</span>
<span class="co">#&gt;    signature_symbol signature_score signature_direction</span>
<span class="co">#&gt; 1              Saa1       -35.29997                   -</span>
<span class="co">#&gt; 3           Sult3a1       -31.75527                   -</span>
<span class="co">#&gt; 4            Isyna1       -29.93255                   -</span>
<span class="co">#&gt; 6             Chil1       -25.41318                   -</span>
<span class="co">#&gt; 8              Saa2       -23.81452                   -</span>
<span class="co">#&gt; 11          Sult1e1       -22.76345                   -</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">signatures</span><span class="op">)</span>
<span class="co">#&gt;     signature_symbol signature_score signature_direction</span>
<span class="co">#&gt; 339              Cfp        7.040545                   +</span>
<span class="co">#&gt; 340        Igkv9-124        7.031580                   +</span>
<span class="co">#&gt; 341           Mtnr1a        7.030133                   +</span>
<span class="co">#&gt; 343          Clec12a        7.024170                   +</span>
<span class="co">#&gt; 344           Angpt2        7.021941                   +</span>
<span class="co">#&gt; 345             Gas2        7.013040                   +</span></code></pre></div>
</div>
<div id="create-omicsignature-object" class="section level3">
<h3 class="hasAnchor">
<a href="#create-omicsignature-object" class="anchor"></a>Create OmicSignature object</h3>
<p>We have everything we need now.<br>
Use <code>OmicSignature$new()</code> to create a new OmicSignature R6 object.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">OmicObj</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/OmicSignature.html">OmicSignature</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>
  metadata <span class="op">=</span> <span class="va">metadata</span>,
  signature <span class="op">=</span> <span class="va">signatures</span>,
  difexp <span class="op">=</span> <span class="va">difexp</span>
<span class="op">)</span>
<span class="co">#&gt;   [Success] OmicSignature object Myc_reduce_mice_liver_24m created.</span></code></pre></div>
<p>Use <code><a href="https://rdrr.io/r/base/print.html">print()</a></code> to see its information:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">OmicObj</span><span class="op">)</span>
<span class="co">#&gt; Signature Object: </span>
<span class="co">#&gt;   Metadata: </span>
<span class="co">#&gt;     animal_strain = C57BL/6 </span>
<span class="co">#&gt;     covariates = none </span>
<span class="co">#&gt;     direction_type = bi-directional </span>
<span class="co">#&gt;     fdr_cutoff = 0.05 </span>
<span class="co">#&gt;     keywords = Myc, KO, longevity </span>
<span class="co">#&gt;     organism = Mus Musculus </span>
<span class="co">#&gt;     phenotype = Myc_reduce </span>
<span class="co">#&gt;     platform = GPL6246 </span>
<span class="co">#&gt;     PMID = 25619689 </span>
<span class="co">#&gt;     sample_type = liver </span>
<span class="co">#&gt;     score_cutoff = 7 </span>
<span class="co">#&gt;     signature_name = Myc_reduce_mice_liver_24m </span>
<span class="co">#&gt;     year = 2015 </span>
<span class="co">#&gt;   signature: </span>
<span class="co">#&gt;     - (152)</span>
<span class="co">#&gt;     + (194)</span>
<span class="co">#&gt;   Differential Expression Data: </span>
<span class="co">#&gt;     27359 x 9</span></code></pre></div>
<p>You can also ask the program to print the messages while creating the OmicSignature Object. By default, <code>print_message</code> is set to be <code>FALSE</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">OmicObj</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/OmicSignature.html">OmicSignature</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>
  metadata <span class="op">=</span> <span class="va">metadata</span>,
  signature <span class="op">=</span> <span class="va">signatures</span>,
  difexp <span class="op">=</span> <span class="va">difexp</span>,
  print_message <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>
<span class="co">#&gt;   difexp: additional columns found: logfc, aveexpr, p_value, b, gene_name. </span>
<span class="co">#&gt;   [Success] difexp matrix is valid. </span>
<span class="co">#&gt;   --Required attributes for metadata: signature_name, organism, platform, direction_type, phenotype, covariates --</span>
<span class="co">#&gt;   [Success] Metadata is saved. </span>
<span class="co">#&gt;   Signature: Checked, signature is bi-directional</span>
<span class="co">#&gt;                          with - (Dn) and + (Up) directions. </span>
<span class="co">#&gt;   [Success] Signature is valid. </span>
<span class="co">#&gt;   [Success] OmicSignature object Myc_reduce_mice_liver_24m created.</span></code></pre></div>
</div>
<div id="extract-new-signatures-from-the-omicsignature-object" class="section level3">
<h3 class="hasAnchor">
<a href="#extract-new-signatures-from-the-omicsignature-object" class="anchor"></a>Extract new signatures from the <code>OmicSignature</code> object</h3>
<p>Previously we created the signature using abs(logfc) &gt; 1 and adj_p &lt; 0.005. Now we can use new criterias to extract new signatures conveniently from the OmicSignature Object.<br>
For example, extract all features with a t-score with absolute value higher than 20 and adj_p less than 0.001:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">OmicObj</span><span class="op">$</span><span class="fu">extract.signature</span><span class="op">(</span><span class="st">"abs(score) &gt; 20; adj_p &lt; 0.001"</span><span class="op">)</span>
<span class="co">#&gt;           symbol     score direction</span>
<span class="co">#&gt; 1           Saa1 -35.29997         -</span>
<span class="co">#&gt; 2           Cpa1  34.62008         +</span>
<span class="co">#&gt; 3        Sult3a1 -31.75527         -</span>
<span class="co">#&gt; 4         Isyna1 -29.93255         -</span>
<span class="co">#&gt; 5           Zg16  28.51562         +</span>
<span class="co">#&gt; 6          Chil1 -25.41318         -</span>
<span class="co">#&gt; 7  2210010C04Rik  24.54446         +</span>
<span class="co">#&gt; 8           Saa2 -23.81452         -</span>
<span class="co">#&gt; 9          Cidec  23.39573         +</span>
<span class="co">#&gt; 10        Amy2a5  22.91294         +</span>
<span class="co">#&gt; 11       Sult1e1 -22.76345         -</span>
<span class="co">#&gt; 12           Pf4  21.71260         +</span>
<span class="co">#&gt; 13          Orm3 -21.25088         -</span>
<span class="co">#&gt; 14          Cpb1  20.98057         +</span></code></pre></div>
</div>
</div>
<div id="write-omicsignature-object-into-a--json-file" class="section level2">
<h2 class="hasAnchor">
<a href="#write-omicsignature-object-into-a--json-file" class="anchor"></a>Write OmicSignature Object into a .json file</h2>
<p>Besides save the OmicSignature object as an <code>.rds</code> file, we can also save it as a <code>.json</code> text file:</p>
<pre><code><a href="../reference/writeJson.html">writeJson(OmicObj, "Myc_reduce_mice_liver_24m.json")</a></code></pre>
<p>We can also read in a OmicSignature object from a <code>.json</code> file.<br>
The following example is from the same Myc reduce study in mice, with muscle and adipose tissue.</p>
<pre><code>OmicObj1 &lt;- readJson(file.path(system.file("extdata", package = "OmicSignature"), "Myc_reduce_mice_muscle_24m_obj.json"))</code></pre>
<p>Thank you! Please contact <a href="mailto:vmli@bu.edu" class="email">vmli@bu.edu</a> if you have any questions.</p>
<p>–</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mengze Li.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
